File: 0020-valid-parentheses/0020-valid-parentheses.java


class Solution {
    public boolean isValid(String s) 
    {
        Stack<Character> st = new Stack<>();
        String pattern = "([{";
        for (int i = 0; i < s.length(); i++) 
        {
            char ch = s.charAt(i);
            if (!st.isEmpty()) 
            {
                char sch = st.peek();
                if(sch == '(' && ch == ')')
                    st.pop();
                else if(sch == '{' && ch == '}')
                    st.pop();
                else if(sch == '[' && ch == ']')
                    st.pop();
                else if(pattern.indexOf(ch) == -1)
                    return false;
                else
                    st.push(ch);
            }
            else
            {
                if(pattern.indexOf(ch) == -1)
                    return false;
                else
                    st.push(s.charAt(i));
            }
                
        }
        return (st.isEmpty())?true:false;
        
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s): # { [ ( ) ] }
        stack = []
        for ch in s:
            if ch == '[' or ch == '(' or ch == '{' :
                stack.append(ch)
            else:
                if len(stack) > 0:
                    top = stack[-1] 
                    if ch == ']' and top == '[':
                        stack.pop()
                    elif ch == '}' and top == '{':
                        stack.pop()
                    elif ch == ')' and top == '(':
                        stack.pop()
                    else:
                        return False
                else:
                    return False
        return len(stack) == 0
        